_posts/2021-03-21-clean-unit-tests-with-mockito.md

```diff
@@ -0,0 +1,681 @@
+
+  void initializeScenarioOne() {
+    // Mockito behaviour declarations
+  }
+
+  void initializeScenarioTwo() {
+    // Mockito behaviour declarations
+  }
+
+  @BeforeEach
+  void setUp() {
+    MockitoAnnotations.openMocks(this);
+    cityService = new CityServiceImpl(cityRepository);
+  }
+
+  @Test
+  void testOne() {
+    initializeScenarioOne();
+    // Test Case One
+  }
+
+  @Test
+  void testTwo() {
+    initializeScenarioTwo();
+    // Test Case Two
+  }
+
+  @Test
+  void testThree() {
+    initializeScenarioOne();
+    // ...
+    initializeScenarioTwo();
+    // ...
+    // Another Test Case
+  }
+}
+```
+
+Extracting the setup in encapsulated methods allows us now to call a certain setup only if needed. We don't have to deal
+with overriding previously declared behaviour.
+
+### Write Test Cases Independently
+
+Don't expect test cases to be always executed in the same order! We write everything that belongs to our test case into 
+the test method, so that a test case can be executed alone in our IDE or all together within our CI!
+
+This mistake often comes together with concatenated setup blocks. If we need to add a new declaration at the end of the
+block, each previous declaration must be executed before we are able to call ours. Or vice versa: if a new declaration
+is inserted at the beginning, causes a shift of all other declarations towards the end. At least now our alarm bell
+should ring, and it's time to reconsider our test case!
+
+### Avoid `Mockito.reset` for Better Unit Tests
+
+Mockito recommends in their documentation to prefer recreation of mocks over resetting them.
+
+> Smart Mockito users hardly use this feature because they know it could be a sign of poor tests.
+> Normally, you don't need to reset your mocks, just create new mocks for each test method.
+
+We better create simple and small test cases than lengthy and over-specified tests. The cause of such tests might be
+testing too much in a single unit test. But let's see an example for this situation.
+
+```java
+  @Test
+  void findAndDelete() throws ElementNotFoundException {
+    City expected = createCity();
+    Mockito.when(cityRepository.find(expected.getId()))
+      .thenReturn(Optional.of(expected));
+    City actual = cityService.find(expected.getId());
+    ReflectionAssert.assertReflectionEquals(expected,actual);
+    cityService.delete(expected);
+    Mockito.verify(cityRepository).delete(expected);
+    Mockito.reset(cityRepository);
+    Mockito.when(cityRepository.find(expected.getId()))
+      .thenReturn(Optional.empty());
+    Assertions.assertThrows(ElementNotFoundException.class,
+      () -> cityService.find(expected.getId()));
+  }
+```
+
+What does this test case?
+
+1. Tries to find a city and an assertion
+2. Deletes a city and a verify
+3. Tries to find the previously city again but expecting an exception.
+
+Such a test case's design is unfortunate. It *tests* too much and could be split in simpler and smaller units.
```

Comment:
Please explain in a sentence what `cityRepository.reset()` does in the code above and why we need it in this example.